src/python/pants/backend/python/dependency_inference/rules.py

```diff
@@ -130,41 +161,35 @@ class InferPythonImportDependencies(InferDependenciesRequest):
     infer_from = PythonSourceField
 
 
-@rule(desc="Inferring Python dependencies by analyzing imports")
-async def infer_python_dependencies_via_imports(
-    request: InferPythonImportDependencies,
-    python_infer_subsystem: PythonInferSubsystem,
-    python_setup: PythonSetup,
-) -> InferredDependencies:
-    if not python_infer_subsystem.imports:
-        return InferredDependencies([])
-
-    _wrapped_tgt = await Get(WrappedTarget, Address, request.sources_field.address)
-    tgt = _wrapped_tgt.target
-    explicitly_provided_deps, parsed_imports = await MultiGet(
-        Get(ExplicitlyProvidedDependencies, DependenciesRequest(tgt[Dependencies])),
-        Get(
-            ParsedPythonImports,
-            ParsePythonImportsRequest(
-                cast(PythonSourceField, request.sources_field),
-                InterpreterConstraints.create_from_targets([tgt], python_setup),
-                string_imports=python_infer_subsystem.string_imports,
-                string_imports_min_dots=python_infer_subsystem.string_imports_min_dots,
-            ),
-        ),
-    )
-
-    resolve = tgt[PythonResolveField].normalized_value(python_setup)
-    owners_per_import = await MultiGet(
-        Get(PythonModuleOwners, PythonModuleOwnersRequest(imported_module, resolve=resolve))
-        for imported_module in parsed_imports
-    )
-
-    merged_result: set[Address] = set()
+def _get_inferred_resource_deps(
+    all_asset_targets: AllAssetTargets,
+    assets: ParsedPythonAssets,
+) -> Iterator[Address]:
+    assets_by_path: Dict[PurePath, Target] = {}
+    for file_tgt in all_asset_targets.files:
+        assets_by_path[PurePath(file_tgt[FileSourceField].file_path)] = file_tgt
+    for resource_tgt in all_asset_targets.resources:
+        path = PurePath(resource_tgt[ResourceSourceField].file_path)
+        assets_by_path[path] = resource_tgt
```

Comment:
This can be slow to create, so I think we want it in a dedicated rule so that we get memoization. Basically `module_mapper.py` for resources. Also...this should probably handle ambiguity similar to Python's `module_mapper.py` - you can't infer if >1 target exposes the same "path"

Reply:
What makes you think it'd be slow? It's pure-python dictionary population so I think it'd be in the noise here.

Reply:
My rough timing in the Pants repo shows it to be less than a second

Reply:
Iterating over `AllTargets` in a repo with 10,000 targets. Users have reported that the Python module mapping creation can take several seconds. Better to memoize that.

Beyond the perf, I do think the ambiguity part is important. We might not want to do the normal warning with ambiguity because of too many false positives, but we at least do want to not infer a dep when ambiguous.

Reply:
Note that you're re-creating the module mapping for every single target: the rule that calls `_get_inferred_asset_deps` is executed every single Python target. This should be memoized through a dedicated rule to create the module mapping.

Reply:
> Note that you're re-creating the module mapping for every single target

:facepalm: Duh...